pkg/controller/stacks/templates/renderphase_controller.go

```diff
@@ -0,0 +1,261 @@
+*/
+
+package templates
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/crossplaneio/crossplane/apis/stacks/v1alpha1"
+	"github.com/crossplaneio/crossplane/pkg/stacks/templates/engines"
+)
+
+// RenderPhaseReconciler reconciles an object which we're watching for a template stack. It handles
+// the render phase of a template stack controller's lifecycle. The render phase looks something like:
+//
+// Inputs:
+// * A stack's behavior configuration
+// * A stack's contents
+// * A CR which triggered the render
+// * Configuration for a resource engine
+//
+// Outputs:
+// * Resources are applied to the cluster
+//
+// Side effects:
+// * A resource engine (such as helm2 or kustomize) is invoked
+//   to create the resources which will be applied to the cluster.
+//   The resource engine's inputs are: configuration specific to the engine,
+//   and the contents of the stack. The outputs are the resources to apply.
+type RenderPhaseReconciler struct {
+	Client     client.Client
+	Log        logr.Logger
+	GVK        *schema.GroupVersionKind
+	EventName  v1alpha1.EventName
+	ConfigName types.NamespacedName
+}
+
+const (
+	renderTimeout = 60 * time.Second
+)
+
+// Reconcile responds to a CR of a particular CRD kind which a template stack is interested in.
+// In most cases, reconciliation means that a resource engine is executed, and the output of the resource
+// engine is applied to the cluster.
+func (r *RenderPhaseReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), renderTimeout)
+	defer cancel()
+
+	// We grab the claim as an unstructured so that we can have the same code handle
+	// arbitrary claim types. The types will be erased by this point, so if a stack
+	// author wants to validate the schema of a claim, they can do it by putting a
+	// schema in the CRD of the claim type so that the claim's schema is validated
+	// at the time that the object is accepted by the api server.
+	i := &unstructured.Unstructured{}
+	i.SetGroupVersionKind(*r.GVK)
+	if err := r.Client.Get(ctx, req.NamespacedName, i); err != nil {
+		if kerrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	// TODO remaining functionality for the render phase:
+	// - Surface the job result somehow (for usability)
+	// - Support for deletion, unless we assume that garbage collection will do it for us
+	return ctrl.Result{}, r.render(ctx, i)
+}
+
+func (r *RenderPhaseReconciler) render(ctx context.Context, claim *unstructured.Unstructured) error {
+	cfg, _ := r.getStackConfiguration(ctx, claim)
+	// TODO check for errors
```

Comment:
> When we create a renderController, we are telling it to watch unstructured types.

This is not entirely accurate, but I addressed the topic in the related thread.

> what happens when you create a 2nd template stack, do the 2 render controllers both fight and try to reconcile the CRDs

This isn't as broad of a concern as in the way it was originally asked, but it is still an issue with the current implementation since there's no mechanism for deduping or coordinating between multiple stacks which are interested in the same CRD type. I'm planning on hardening this part of the controller in the future.